src/smc-webapp/jupyter/parsing.ts

```diff
@@ -35,3 +36,25 @@ function last_style(code: string, mode = "python") {
   CodeMirror.runMode(code, mode, (_, s) => (style = s));
   return style;
 }
+
+// calling escape and then unescape should be idempotent
+// python is inspired by https://github.com/drillan/jupyter-black/blob/master/jupyter-black.js
+export function process_magics(
```

Comment:
Is it really possible to solve this problem using regexps?  E.g., What happens with this valid Python code that has no magics in it (it's just `3%2` but on two lines)?
```
3\
%2
```